lib/SIL/Verifier/ReborrowVerifier.cpp

```diff
@@ -0,0 +1,95 @@
+//===--- ReborrowVerifier.cpp ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-reborrow-checker"
+
+#include "ReborrowVerifierPrivate.h"
+
+using namespace swift;
+
+bool ReborrowVerifier::verifyReborrowLifetime(SILPhiArgument *phiArg,
+                                              SILValue baseVal) {
+  SmallPtrSet<SILBasicBlock *, 4> visitedBlocks;
+  bool result = false;
+  SmallVector<Operand *, 4> phiArgUses(phiArg->getUses().begin(),
+                                       phiArg->getUses().end());
+
+  // Verify whethre the guaranteed phi arg lies within the lifetime of the base
+  // value.
+  LinearLifetimeChecker checker(visitedBlocks, deadEndBlocks);
+  // newErrorBuilder is consumed at the end of the checkValue function.
+  // Copy initial state from errorBuilder everytime
+  LinearLifetimeChecker::ErrorBuilder newErrorBuilder = errorBuilder;
+  SmallVector<Operand *, 4> baseValConsumingUses(baseVal->getConsumingUses());
+  // If the baseValue has no consuming uses, there is nothing more to verify
+  if (baseValConsumingUses.empty())
+    return false;
+  auto linearLifetimeResult = checker.checkValue(baseVal, baseValConsumingUses,
+                                                 phiArgUses, newErrorBuilder);
+  result |= linearLifetimeResult.getFoundError();
+  return result;
+}
+
+void ReborrowVerifier::verifyReborrows(BorrowingOperand initialScopedOperand,
+                                       SILValue value) {
+  SmallVector<std::tuple<Operand *, SILValue>, 4> worklist;
+  // Initialize the worklist with borrow lifetime ending uses
+  initialScopedOperand.visitLocalEndScopeInstructions([&](Operand *op) {
+    worklist.push_back({op, value});
+  });
+
+  while (!worklist.empty()) {
+    Operand *borrowLifetimeEndOp;
+    SILValue baseVal;
+    std::tie(borrowLifetimeEndOp, baseVal) = worklist.pop_back_val();
+    auto *borrowLifetimeEndUser = borrowLifetimeEndOp->getUser();
+
+    if (!isReborrowInstruction(borrowLifetimeEndUser)) {
+      continue;
+    }
+
+    if (isVisitedOp(borrowLifetimeEndOp, baseVal))
+      continue;
+
+    // Process reborrow
```

Comment:
I would add a TODO here that says if we add more reborrow instructions this should be abstracted onto a ReborrowingOperand ADT type so we can abstract over terminator vs non-terminator. I am fine doing that latter.